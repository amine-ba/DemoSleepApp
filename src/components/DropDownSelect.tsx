import React from "react";
import {
  CheckIcon,
  FormControl,
  Select,
  WarningOutlineIcon,
} from "native-base";

import { locals } from "@consts";

export interface DropDownSelect {
  items: SelectItem[];
  value: string;
  onValueChange: (value: string) => void;
  title?: string;
  placeholder?: string;
  validate?: (value: string) => string | null;
}

export const DropDownSelect = ({
  items,
  value,
  onValueChange,
  title,
  placeholder,
  validate,
}: DropDownSelect) => {
  const errorMessage = validate?.(value);

  return (
    <FormControl isInvalid={!!errorMessage}>
      <FormControl.Label mb="3">{title}</FormControl.Label>

      <Select
        selectedValue={value}
        minWidth="200"
        accessibilityLabel={placeholder}
        placeholder={placeholder}
        _selectedItem={{
          bg: "teal.600",
          endIcon: <CheckIcon size="5" />,
        }}
        mt={1}
        onValueChange={onValueChange}
      >
        {items.map(({ label, value }) => (
          <Select.Item key={label} label={label} value={value} />
        ))}
      </Select>

      <FormControl.ErrorMessage leftIcon={<WarningOutlineIcon size="xs" />}>
        {errorMessage || locals["dropdown.select.default.error"]}
      </FormControl.ErrorMessage>
    </FormControl>
  );
};

export default DropDownSelect;
