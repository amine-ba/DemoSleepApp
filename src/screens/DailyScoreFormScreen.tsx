import React, { useCallback, useLayoutEffect, useState } from "react";

import {
  CheckIcon,
  Button,
  Heading,
  FormControl,
  WarningOutlineIcon,
} from "native-base";

import theme from "@shared/theme";
import { FormContainer, SpacingContainer } from "@shared/styles";

import DropDownSelect from "@components/DropDownSelect";
import DisplayScoreScreen from "@screens/DisplayScoreScreen";

import { RequestStatus, useMutate } from "@api";

import { TextContent, DurationSelectOptions } from "@consts";
import { useScore } from "hooks/useScore";

export const DailyScoreFormScreen = () => {
  const [durationInBed, setDurationInBed] = useState<string>("");
  const [durationAsleep, setDurationAsleep] = useState<string>("");
  const [errorMessage, setErrorMessage] = useState<string | null>();

  const [displayScore, setDisplayScore] = useState<boolean>(false);

  const score = useScore({ durationInBed, durationAsleep });

  const { mutate, status, error } = useMutate();

  const reset = () => {
    setDurationInBed("");
    setDurationAsleep("");
    setErrorMessage(null);
    setDisplayScore(false);
  };

  const canSubmit = () => {
    return (
      !!durationInBed &&
      !!durationAsleep &&
      Number(durationInBed) > Number(durationAsleep)
    );
  };

  const onSubmit = () => {
    mutate("localhost:8080/scores", score);
  };

  const validate = useCallback(() => {
    return !durationAsleep || !durationInBed || canSubmit();
  }, [durationAsleep, durationInBed]);

  useLayoutEffect(() => {
    if (status === RequestStatus.initial || status === RequestStatus.loading)
      return;

    if (status === RequestStatus.success) {
      setDisplayScore(true);
    } else {
      setErrorMessage(error);
    }
  }, [status]);

  if (displayScore && score)
    return <DisplayScoreScreen reset={reset} score={score} />;

  return (
    <FormContainer safeAreaTop>
      <SpacingContainer mBottom={50}>
        <Heading>{TextContent.scoreForm.title}</Heading>
      </SpacingContainer>

      <SpacingContainer mVertical={30}>
        <DropDownSelect
          value={durationInBed}
          items={DurationSelectOptions}
          onValueChange={(value: string) => {
            if (errorMessage) setErrorMessage(null);
            setDurationInBed(value);
          }}
          title={TextContent.scoreForm.durationInBed.title}
          placeholder={TextContent.scoreForm.durationInBed.placeholder}
          validate={() => {
            if (validate()) return null;
            else return TextContent.scoreForm.durationInBed.error;
          }}
        />
      </SpacingContainer>

      <DropDownSelect
        value={durationAsleep}
        items={DurationSelectOptions}
        onValueChange={(value: string) => {
          if (errorMessage) setErrorMessage(null);
          setDurationAsleep(value);
        }}
        title={TextContent.scoreForm.durationAsleep.title}
        placeholder={TextContent.scoreForm.durationAsleep.placeholder}
        validate={() => {
          if (validate()) return null;
          else return TextContent.scoreForm.durationAsleep.error;
        }}
      />

      <SpacingContainer mVertical={20}>
        <FormControl isInvalid={!!errorMessage}>
          <FormControl.ErrorMessage leftIcon={<WarningOutlineIcon size="xs" />}>
            {errorMessage}
          </FormControl.ErrorMessage>
        </FormControl>
      </SpacingContainer>

      <SpacingContainer mTop={60}>
        <Button
          testID="score-form-submit-button"
          isLoading={status === RequestStatus.loading}
          endIcon={<CheckIcon size="xs" />}
          isDisabled={!canSubmit()}
          colorScheme={theme.colors.primary.green}
          onPress={onSubmit}
        >
          {TextContent.submit}
        </Button>
      </SpacingContainer>
    </FormContainer>
  );
};

export default DailyScoreFormScreen;
