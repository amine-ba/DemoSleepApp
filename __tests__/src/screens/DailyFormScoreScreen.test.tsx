import React from "react";
import { render, fireEvent } from "@testing-library/react-native";

import DailyScoreFormScreen from "@screens/DailyScoreFormScreen";

import { NativeBaseProvider } from "native-base";
import { DropDownSelectProps } from "@components/DropDownSelect";
import { act } from "react-test-renderer";

const inset = {
  frame: { x: 0, y: 0, width: 0, height: 0 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
};

const TestComponent = () => (
  <NativeBaseProvider initialWindowMetrics={inset}>
    <DailyScoreFormScreen />
  </NativeBaseProvider>
);

const MockMutate = jest.fn();
const ScoreMock = 70;

jest.mock("../../../src/hooks/useScore/index.ts", () => ({
  useScore: (_props: any) => ScoreMock,
}));

jest.mock("../../../src/hooks/useMutate/index.ts", () => ({
  ...jest.requireActual("../../../src/hooks/useMutate/index.ts"),
  useMutate: () => ({
    mutate: MockMutate,
    status: "finished",
    error: null,
  }),
}));

const durationInBedMock = "180";
const durationAsleepMock = "90";

let setDurationInBedMockFuntion = jest.fn();
let setDurationAsleepMockFuntion = jest.fn();

jest.mock(
  "../../../src/components/DropDownSelect",
  () =>
    ({ testID, onValueChange }: DropDownSelectProps) => {
      if ((testID = "durationInBedSelect"))
        setDurationInBedMockFuntion = jest.fn((value: string) => {
          onValueChange(value);
        });

      if ((testID = "durationInBedSelect"))
        setDurationAsleepMockFuntion = jest.fn((value: string) => {
          onValueChange(value);
        });

      return <></>;
    }
);

describe(" <DailyScoreFormScreen /> Component ", () => {
  it(" Should Successfully Rendred ", () => {
    const json = render(<TestComponent />).toJSON();

    expect(json).toMatchSnapshot();
  });

  it(" Should Save the correct Score ", async () => {
    const { findByTestId } = render(<TestComponent />);
    
    const submitButtonNode = await findByTestId("save-score-button");

    act(() => {
      setDurationInBedMockFuntion(durationInBedMock);
      setDurationAsleepMockFuntion(durationAsleepMock);

      fireEvent.press(submitButtonNode);

      expect(MockMutate).not.toBeCalledWith("", 0);
    });
  });
});
