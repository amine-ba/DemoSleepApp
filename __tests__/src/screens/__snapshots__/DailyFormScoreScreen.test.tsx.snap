// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <DailyScoreFormScreen /> Component   Should Successfully Rendred  1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    className="sc-bdvvtL sc-gsDKAQ  giHPxp"
    dataSet={Object {}}
    style={Object {}}
  >
    <View
      className="sc-dkPtRN ipVPo"
      dataSet={Object {}}
      mBottom={50}
      style={Object {}}
    >
      <Text
        dataSet={Object {}}
        style={
          Object {
            "backgroundColor": undefined,
            "color": "#404040",
            "fontFamily": undefined,
            "fontSize": null,
            "fontStyle": "normal",
            "fontWeight": "700",
            "letterSpacing": 0,
            "lineHeight": 30,
            "textDecorationLine": undefined,
          }
        }
      >
        How Well You Slept ?
      </Text>
    </View>
    <View
      className="sc-dkPtRN gJaBSN"
      dataSet={Object {}}
      mVertical={30}
      style={Object {}}
    />
    <View
      className="sc-dkPtRN hPqa-Dd"
      dataSet={Object {}}
      mVertical={20}
      style={Object {}}
    >
      <View
        dataSet={Object {}}
        isInvalid={false}
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </View>
    <View
      className="sc-dkPtRN ceYxvD"
      dataSet={Object {}}
      mTop={60}
      style={Object {}}
    >
      <View
        _text={
          Object {
            "color": "darkText",
          }
        }
        accessibilityRole="button"
        accessible={true}
        colorScheme="green"
        dataSet={Object {}}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onHoverIn={[Function]}
        onHoverOut={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d4d4d4",
            "borderRadius": 4,
            "flexDirection": "row",
            "justifyContent": "center",
            "opacity": 0.5,
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
          }
        }
        testID="save-score-button"
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#000000",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "500",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Sumbmit
            </Text>
          </View>
          <View
            dataSet={Object {}}
            style={
              Object {
                "width": 8,
              }
            }
          />
          <RNSVGSvgView
            accessibilityRole="image"
            align="xMidYMid"
            bbHeight={16}
            bbWidth={16}
            color={4278190080}
            dataSet={Object {}}
            focusable={false}
            meetOrSlice={0}
            minX={0}
            minY={0}
            stroke=""
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "color": "#000000",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "height": 16,
                  "width": 16,
                },
                Object {
                  "flex": 0,
                  "height": 16,
                  "width": 16,
                },
              ]
            }
            tintColor={4278190080}
            vbHeight={14}
            vbWidth={14}
          >
            <RNSVGGroup
              font={
                Object {
                  "fontSize": 14,
                  "fontWeight": "500",
                }
              }
              propList={
                Array [
                  "stroke",
                ]
              }
              stroke={null}
            >
              <RNSVGGroup>
                <RNSVGGroup
                  fill={
                    Array [
                      2,
                    ]
                  }
                  propList={
                    Array [
                      "fill",
                      "stroke",
                    ]
                  }
                  stroke={null}
                >
                  <RNSVGPath
                    d="M5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039z"
                  />
                </RNSVGGroup>
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
